rp2040-hal-examples/src/bin/spi_eh_bus.rs

```diff
@@ -0,0 +1,197 @@
+    spi: SPI,
+}
+
+/// When things go wrong, we could return Error(()) but that wouldn't help anyone troubleshoot
+/// so we'll create an error type with additional info to return.
+#[derive(Copy, Clone, Debug)]
+enum MyError<SPI> {
+    Spi(SPI),
+    // Add other errors for your driver here.
+}
+
+/// Implementation of the business logic for the remote Spi IC
+impl<SPI> MySpiDriver<SPI>
+where
+    SPI: SpiDevice,
+{
+    /// Construct a new instance of Spi device driver
+    pub fn new(spi: SPI) -> Self {
+        Self { spi }
+    }
+
+    /// Our hypothetical Spi device has a register at 0x20, that accepts a u8 value
+    pub fn set_value(&mut self, value: u8) -> Result<(), MyError<SPI::Error>> {
+        self.spi
+            .transaction(&mut [Operation::Write(&[0x20, value])])
+            .map_err(MyError::Spi)?;
+
+        Ok(())
+    }
+
+    /// Our hypothetical Spi device has a register at 0x90, that we can read a 2 byte value from
+    pub fn get_value(&mut self) -> Result<[u8; 2], MyError<SPI::Error>> {
+        let mut buf = [0; 2];
+        self.spi
+            .transaction(&mut [Operation::Write(&[0x90]), Operation::Read(&mut buf)])
+            .map_err(MyError::Spi)?;
+
+        Ok(buf)
+    }
+}
+
+/// Entry point to our bare-metal application.
+///
+/// The `#[rp2040_hal::entry]` macro ensures the Cortex-M start-up code calls this function
+/// as soon as all global variables and the spinlock are initialised.
+///
+/// The function configures the RP2040 peripherals, then performs some example
+/// SPI transactions, then goes to sleep.
+#[rp2040_hal::entry]
+fn main() -> ! {
+    // Grab our singleton objects
+    let mut pac = pac::Peripherals::take().unwrap();
+
+    // Set up the watchdog driver - needed by the clock setup code
+    let mut watchdog = hal::Watchdog::new(pac.WATCHDOG);
+
+    // Configure the clocks
+    let clocks = hal::clocks::init_clocks_and_plls(
+        XTAL_FREQ_HZ,
+        pac.XOSC,
+        pac.CLOCKS,
+        pac.PLL_SYS,
+        pac.PLL_USB,
+        &mut pac.RESETS,
+        &mut watchdog,
+    )
+    .unwrap();
+
+    // The single-cycle I/O block controls our GPIO pins
+    let sio = hal::Sio::new(pac.SIO);
+
+    // Set the pins to their default state
+    let pins = hal::gpio::Pins::new(
+        pac.IO_BANK0,
+        pac.PADS_BANK0,
+        sio.gpio_bank0,
+        &mut pac.RESETS,
+    );
+
+    let timer = rp2040_hal::Timer::new(pac.TIMER, &mut pac.RESETS, &clocks);
+
+    let uart_pins = (
+        // UART TX (characters sent from RP2040) on pin 1 (GPIO0)
+        pins.gpio0.into_function(),
+        // UART RX (characters received by RP2040) on pin 2 (GPIO1)
+        pins.gpio1.into_function(),
+    );
+
+    // Set up a uart so we can print out the values from our Spi peripheral
+    let mut uart = hal::uart::UartPeripheral::new(pac.UART0, uart_pins, &mut pac.RESETS)
+        .enable(
+            UartConfig::new(9600.Hz(), DataBits::Eight, None, StopBits::One),
```

Comment:
Half our examples run at 9600 and half at 115200. I prefer the latter.

Reply:
same.